### PR TITLE
Issue 500 - fixed reporting of supported opcodes for optable(DOS,LIST)

### PR DESCRIPTION
**Pull request by abekornelis**
- Issue 500 - fixed reporting of supported opcodes for optable(DOS,LIST)
- Fixed WTO macro using an instruction that is not supported under optable(DOS)

### CODE REVIEW

**Comment by abekornelis** [comment]
@adelosa could you please provide some help getting the new test incorporated into the groovy stuff?

File: mac/ZRTOPTBL.CPY [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+* Program exit, returncode in field RETCODE
+*
+***********************************************************************
+EXIT     DS    0H                      * Workarea handling
+         XR    R15,R15                 * Reset for return code
+         IC    R15,RETCODE             * Pick up return code
+         L     R13,4(,R13)             * Get ptr to prev savearea
+         L     R14,12(,R13)            * Retrieve return address
+*******  L     R15,16(,R13)            * Return code already in R15
+         LM    R0,R12,12(R13)          * Reset all other registers
```
Comment (jyganci) [inline mac/ZRTOPTBL.CPY]: Should be "LM    R0,R12,20(R13) ..."

File: mac/ZRTOPTBL.CPY [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+         L     R5,=AL4(Z390TBL)        * Point entries buffer
+
+*
+* Read record, at end set eof indicator
+GETZ390_READ   DS 0H
+         LA    R14,GETZ390_READ2       * counter z390 issue #525
+         GET   Z390PRN,Z390REC         * Retrieve record
+GETZ390_READ2  DS 0H                   * Resume point after EOF
+         LTR   R15,R15                 * Check return code
+         BNZ   ERROR03                 * On error exit
```
Comment (jyganci) [inline mac/ZRTOPTBL.CPY]: A GET invocation does not put a return code in R15.

File: mac/ZRTOPTBL.CPY [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+         BE    GETZ390_END             * Okay - we're done
+
+         CLI   Z390EOF,X'10'           * opcode table detected?
+         BE    GETZ390_PROCESS         * Yes: process input record
+
+*
+* Opcode table not yet started: scan for marker
+         CLC   MARKER,Z390REC+1        * Record has marker?
+         BNE   GETZ390_READ            * No: ignore record
+         OI    Z390EOF,X'10'           * Indicate marker found
```
Comment (jyganci) [inline mac/ZRTOPTBL.CPY]: Used "CLI   Z390EOF,X'10'" above so should use MVI, not OI, here.

File: mac/ZRTOPTBL.CPY [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+         L     R5,=AL4(HLASMTBL)       * Point entries buffer
+
+*
+* Read record, at end set eof indicator
+GETHLASM_READ   DS 0H
+         LA    R14,GETHLASM_READ2      * counter z390 issue #525
+         GET   HLASMPRN,HLASMREC       * Retrieve record
+GETHLASM_READ2  DS 0H                  * Resume point after EOF
+         LTR   R15,R15                 * Check return code
+         BNZ   ERROR04                 * On error exit
```
Comment (jyganci) [inline mac/ZRTOPTBL.CPY]: Same comment as for z390 reads.

File: mac/ZRTOPTBL.CPY [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+         BE    GETHLASM_END            * Okay - we're done
+
+         CLI   HLASMEOF,X'10'          * opcode table detected?
+         BE    GETHLASM_PROCESS        * Yes: process input record
+
+*
+* Opcode table not yet started: scan for marker
+         CLC   MARKER,HLASMREC+1       * Record has marker?
+         BNE   GETHLASM_READ           * No: ignore record
+         OI    HLASMEOF,X'10'          * Indicate marker found
```
Comment (jyganci) [inline mac/ZRTOPTBL.CPY]: Same comment as for Z390EOF.

File: mac/ZRTOPTBL.CPY (hunk 1 above)
Comment (abekornelis) [inline mac/ZRTOPTBL.CPY:44]: This is where John noticed the offset for the LM is incorrect.

**Review by abekornelis — COMMENTED** [review]
I added a single comment in the location where John detected an incorrect offset. I'm wondering if it will show up now.

File: src/mz390.java [hunk 6]
```diff
@@ -8492,7 +8493,7 @@ private void init_gbl_sys(){
 		gbl_setc[gbl_sysm_sev_index] = "000"; // RPI 898
 		add_gbl_sys("&SYSOPT_DBCS",var_setb_type); // set false indicating no DBCS support
 		add_gbl_sys("&SYSOPT_OPTABLE",var_setc_type); // "z390" opcode table name
-		gbl_setc[tot_gbl_setc-1] = "z390";
+        gbl_setc[tot_gbl_setc-1] = tz390.opt_optable; // #527
```
Comment (jyganci) [inline src/mz390.java]: The next line, while not part of your changes, is messed up. It starts with a semicolon. If you feel like it, clean it up by removing the leading semicolon and align the remaining text with the surrounding lines. Just a suggestion.

File: src/az390.java [hunk 7]
```diff
@@ -1889,383 +1927,441 @@ else if (tz390.op_trace_type[index]==142)
...
+                    my_format="SS";                                                    // #500
                     if (tz390.op_name[index].equals("PKU"))
-                       {entry=entry+"   D1(B1),D2(L2,B2)";
+                       {my_operands="D1(B1),D2(L2,B2)";                                // #500
                         }
                     else if (tz390.op_name[index].equals("MVCK")
                     ||  tz390.op_name[index].equals("MVCP")
                     ||  tz390.op_name[index].equals("MVCS"))
-                       {entry=entry+"   D1(R1,B1),D2(B2),R3";
+                       {entry=my_operands="D1(R1,B1),D2(B2),R3";                       // #500
```
Comment (jyganci) [inline src/az390.java]: Question only. Did you really mean to set both entry and my_operands?

File: mac/ZRTOPTBL.CPY [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+HLASMMPT DS    AL4                     * Ptr to empty entry (=end)
+
+Z390EOF  DC    XL1'00'
+HLASMEOF DC    XL1'00'
+RETCODE  DC    XL1'00'                 * Return code from program
+Z390REC  DS    CL(&Z390LEN)
+HLASMREC DS    CL(&HLASMLEN)
+
+         DS    0D
+Z390TBL  DS    5000CL(&ENTRYLEN)       * room for 5000 opcodes
```
Comment (jyganci) [inline mac/ZRTOPTBL.CPY]: Suggestion only. Consider using an equated value in place of the 5000 here and in the HLASMTBL.declaration.

File: mac/ZRTOPTBL.CPY [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+         L    R14,COMPARE_RET          * Get return address
+         BR   R14                      * Return
+
+***********************************************************************
+*
+* EOF routines
+*
+***********************************************************************
+Z390END  DS   0H
+         OI   Z390EOF,X'FF'            * switch eof indicator on
```
Comment (jyganci) [inline mac/ZRTOPTBL.CPY]: Suggestion only. Change OI to MVI. Also at HLASMEND.

File: mac/ZRTOPTBL.CPY [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+         BR   R14
+
+***********************************************************************
+*
+* Error routines
+*
+***********************************************************************
+ERROR01  DS    0H
+         WTO   'Error: Open failed',ROUTCDE=11,DESC=(6)
+         MVI   RETCODE,X'08'
```
Comment (jyganci) [inline mac/ZRTOPTBL.CPY]: Suggestion only. For consistency with other RETCODE settings, use 8 in place of X'08'.

File: mac/ZRTOPTBL.CPY [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+         BH    ERROR07                  * Missing z390 opcode
+         CLC   0(&ENTRYLEN.,R4),0(R6)   * Compare entire entry
+         BNE   ERROR08                  * Report incorrec definition
+         LA    R4,&ENTRYLEN.(,R4)       * Bump z390 entry pointer
+         LA    R6,&ENTRYLEN.(,R6)       * Bump hlasm entry pointer
+COMPARE_NEXT   DS 0H                    * advance to next entry
+         CLR   R4,R5                    * z390 table exhausted?
+         BL    COMPARE_LOOP             * no: compare next entry
+         CLR   R6,R7                    * hlasm table exhausted?
+         BL    COMPARE_LOOP             * no: compare next entry
```
Comment (jyganci) [inline mac/ZRTOPTBL.CPY:205]: Observation.

If the z390 table is larger than the HLASM table, then the first "BL" might be using an invalid R6???
If the HLASM table is larger than the z390 table, then the second "BL" might be using an invalid R4???

Your artificial "999999999" entries at the ends of the tables may prevent these possibilities.

File: src/mz390.java [hunk 12]
```diff
@@ -8492,8 +8493,8 @@ private void init_gbl_sys(){
 		gbl_setc[gbl_sysm_sev_index] = "000"; // RPI 898
 		add_gbl_sys("&SYSOPT_DBCS",var_setb_type); // set false indicating no DBCS support
 		add_gbl_sys("&SYSOPT_OPTABLE",var_setc_type); // "z390" opcode table name
-		gbl_setc[tot_gbl_setc-1] = "z390";
-		;	add_gbl_sys("&SYSOPT_RENT",var_setb_type); // set true if RENT option specified
+        gbl_setc[tot_gbl_setc-1] = tz390.opt_optable; // #527
```
Comment (jyganci) [inline src/mz390.java]: New line 8496 is now too far to the left.

**Review by jyganci — APPROVED** [review]
Final quick scan of changes. Appears to be okay.

**Pull request merged**